### PR TITLE
Detection of significant evolution of an indicator

### PR DESCRIPTION
**Pull request by JanJereczek**
evolution_metrics.jl:
- indicator_evolution() computes indicator and its evolution for original timeseries and surrogates
- chapter "Trend metrics" has been initialized with ridge_regression. Still to come: chapter with metrics for change-point detetction

significance.jl:
- measure_significance() compares results from original timeseries and surrogates.
- chapter "Percentile metrics" has been initialized with various measures of the significance defined wrt to percentiles.

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -4,6 +4,9 @@ authors = ["Jan Swierczek-Jereczek <jan.jereczek@gmail.com>", "George Datseris <
 version = "0.1.0"
 
 [deps]
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
+StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
```
Comment (Datseris) [inline Project.toml:9]: When you add packages to the [deps] you also need to add versions to the [compat]

File: src/evolution_metrics.jl [hunk 2]
```diff
@@ -0,0 +1,151 @@
+struct IndicatorEvolutionResults{T<:AbstractFloat}
+    t_indicator::AbstractVector{T}
+    x_indicator::AbstractVector{T}
+    t_evolution::AbstractVector{T}
+    x_evolution::AbstractVector{T}
+    surr_x_evolution::AbstractMatrix{T}
```
Comment (Datseris) [inline src/evolution_metrics.jl:6]: All of these fields are abstract types (bad for performance). Why thought? Is there a specific reason we don't just have `Vector{T}`? We are computing these quantities, right?

File: src/evolution_metrics.jl [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+end
+
+"""
+    indicator_evolution(t, x, indicator, n_surr, surr_method, evolution_metric,
+                        wv_indicator_width, wv_indicator_stride,
+                        wv_evolution_width, wv_evolution_stride)
```
Comment (Datseris) [inline src/evolution_metrics.jl:12]: Okay, here almost everything should be a keyword argument. Only `t, x, indicator` should be arguments, and everything else a keyword with sensible defaults.

Comment (Datseris) [reply src/evolution_metrics.jl:12]: Additionally, `t` should be an _optional_ argument. If not given, `1:length(x)` is taken as the time vector.

File: src/evolution_metrics.jl [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+Computes the `evolution_metric` of an `indicator` for a timeseries (`t`, `x`) and
+`n_surr` surrogates generated by `surr_method`. The windowing parameters for the
+computation of the `indicator` are `wv_indicator_width` and `wv_indicator_stride`.
+The windowing parameters for the computation of the `evolution_metric`, are
+`wv_evolution_width` and `wv_evolution_stride`.
+
+Future dev: take vector of indicators such that surrogates need to be generated only once.
```
Comment (Datseris) [inline src/evolution_metrics.jl:20]: This is easy to do via multiple dispatch, where `indicator` is a `Vector{Function}`.

File: src/evolution_metrics.jl [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    wv_indicator = WindowViewer(x, wv_indicator_width, wv_indicator_stride)
+    t_indicator = t[wv_indicator.strided_indices]
+    x_indicator = map(indicator, wv_indicator)
+
+    wv_evolution = WindowViewer(x_indicator, wv_evolution_width, wv_evolution_stride)
+    t_evolution = t_indicator[wv_evolution.strided_indices]
+    x_evolution = map(evolution_metric, wv_evolution)
+
+    sgen = surrogenerator(x, surr_method)
```
Comment (Datseris) [inline src/evolution_metrics.jl:43]: You should specify an `rng` here.

File: src/evolution_metrics.jl [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+Computes the `evolution_metric` of an `indicator` for a timeseries (`t`, `x`) and
+`n_surr` surrogates generated by `surr_method`. The windowing parameters for the
+computation of the `indicator` are `wv_indicator_width` and `wv_indicator_stride`.
+The windowing parameters for the computation of the `evolution_metric`, are
+`wv_evolution_width` and `wv_evolution_stride`.
+
+Future dev: take vector of indicators such that surrogates need to be generated only once.
+"""
+@inline function indicator_evolution(
```
Comment (Datseris) [inline src/evolution_metrics.jl:22]: You shouldn't inline this function. It isn't something small that can be inlined and it wont be called by deep calls from other functions.

File: src/evolution_metrics.jl [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+    ridge(t, x; lambda)
+
+Perform ridge regression of `x` over `t` with regularization parameter `lambda`.
+Return vector containing slope and offset.
+If `lambda = 0`, linear regression is recovered (default case).
+For more information, visit: https://en.wikipedia.org/wiki/Ridge_regression
+"""
+@inline function ridge(
+    t::AbstractVector{T},
```
Comment (Datseris) [inline src/evolution_metrics.jl:76]: Here you should make `t` an optional argument that becomes `eltype(x).(1:length(x))` if not given.

File: src/evolution_metrics.jl [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+    lambda::T = T(0),
+) where {T<:Real}
+    return ridge(x, y, lambda = lambda)[1]
+end
+
+@inline function ridge_slope(
+    y::AbstractVector{T},
+) where {T<:Real}
+    return ridge(T.(eachindex(y)), y)[1]
+end
```
Comment (Datseris) [inline src/evolution_metrics.jl:117]: All of this stuff can be trivially done as `ridge_slope(args...; kwargs...) = ridge(args...; kwargs...)[2]`.

File: src/significance.jl [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+"""
+
+    percentile_idx(s; p, symmetric)
+
+If `symmetric == false`, return indices of `0`-th and `p*100`-th percentile of `s`.
+If `symmetric == true` and `p=0.95`, return indices of `2.5`-th and `97.5`-th percentile.
+"""
+@inline function percentile_idx(
+    s::AbstractVector{T};
+    p::T = T(0.95),
```
Comment (Datseris) [inline src/significance.jl:46]: Why should the percentile match type with `x`? I think it shouldn't. In fact, there is no real reason to specify the type of `p` here, and just return `T(i_lo), T(i_hi)` at the end.

**Review by Datseris — COMMENTED** [review]
Jan, this is looking great!

I've added several comments in my review. The most important comment however is the following one: 

It will do us great good if we have a much more functional approach to the codebase. You can separate your main function `indicator_evolution` so that it calls smaller functions.

- [x] We need one function that calculates indicator over a sliding window. We have this, it is just `map(indicator, WindowViewer(x, ...))`.
- [ ] We need one function that calculates some kind of slope / regression / metaanalysis of the time evolution of the indicator. You have the code for it, you just haven't made it into a dedicated function.
- [ ] The above two functions are also called for the surrogates. This immediatelly solves your prolem of doing the same computations for a different indicator but without repeating surrogates: you _first_ compute the surrogate, and for each surrogate you call _exactly_ the same function as you call for the original timeseries.

We also have to discuss namings a bit... `indicator_evolution` isn't really what this function does. The indicator evolution is just the call `map(indicator, WindowViewer(x))`.

**Comment by JanJereczek** [comment]
Thanks a lot for your review George! I am already noticing some benefits of our work for my own coding habits :) Most of what you say is crystal clear apart from two things:

Bullet point 2: Do you mean wrapping the following lines into a function?
```julia
wv_surr_indicator = WindowViewer(s, wv_indicator_width, wv_indicator_stride)
surr_x_indicator = map(indicator, wv_surr_indicator)
wv_surr_evolution = WindowViewer(surr_x_indicator, wv_evolution_width, wv_evolution_stride)
surr_evolution_ts = map(evolution_metric, wv_surr_evolution)
```

Bullet point 3: Do you mean generating the surrogates before the loop and basically saving them? I had thought of providing a vector of indicators (instead of just one as it is the case right now) and make the inner-most loop go over them. Maybe even provide a vector of evolution metrics if we don't want to use the same for all indicators. This way I wouldn't need to store the surrogates. Something like:

```julia
# Just an example of how these vectors would look like
indicators = [var, ar1]
evolution_metrics = [ridge_slope, corr_kendall]

# here some code is missing but we get the point
for i in 1:n_surro
    for k in 1:n_indicators
         s = sgen()
         surr_evolutions[i,:,k] = compute_indicator_evolution(s, indicators[k], evolution_metrics[k])
    end
end
```

Naming: I had the feeling that `indicator_evolution()` was a sensible name as we compute the evolution (either trend or change point - we need a generic name to describe both options) of an indicator (the call of `map` that you mention that gives the time-series but not the evolution metric yet!). I am open for suggestions though!

All the comments regarding code make perfect sense and I will embed them asap :)

**Comment by Datseris** [comment]
What's the case of using `corr_kendall`? The first case with the ridge is clear: we are computing the slope of e.g., AR1 coefficient. And we want to see a significant increase of slope. What is the correlation coefficient supposed to yield? And the correlation between what? The meta-analysis descriptor (what you call evolution metrics) are functions of 1 argument: a timeseries. What is the Kendall coefficient computing here?

**Comment by Datseris** [comment]
> Bullet point 3: Do you mean generating the surrogates before the loop and basically saving them? I had thought of providing a vector of indicators (instead of just one as it is the case right now) and make the inner-most loop go over them. Maybe even provide a vector of evolution metrics if we don't want to use the same for all indicators. This way I wouldn't need to store the surrogates. Something like:

Yes, exactly like that.

**Comment by Datseris** [comment]
> Bullet point 2: Do you mean wrapping the following lines into a function?

Well yes. The function doesn't care if the input is a surrogate or not; it does the same thing for both the real signal and the surrogate anyways. So you define a function once and call it twice: for the signal and the surrogate.

**Comment by JanJereczek** [comment]
In many publications, the kendall-tau correlation is used to measure the trend. It yields 1 for monotonously increasing time-series and -1 for monotonously decreasing one (and any value in-between if there is no monotonous behaviour). It is part of `StatsBase.jl` (https://juliastats.org/StatsBase.jl/stable/ranking/#StatsBase.corkendall) and for now I just used it as a placeholder. I wanted to illustrate that we might want to use the evolution metric n°1 to measure the trend of indicator A, and evolution metric n°2 to measure the trend of indicator B.

**Comment by JanJereczek** [comment]
Note that `corrkendall(x, y=x)` is by default a 1-argument function following that formula:

```math
\tau=\frac{2}{n(n-1)} \sum \operatorname{sgn}\left(x_i-x_j\right) \operatorname{sgn}\left(y_i-y_j\right)
```

**Comment by Datseris** [comment]
okj sounds good. The "metaanalysis metric" (which is what you call "evolution metric") can be either a function or a vector of functions of equal length to the vector of indicators.

**Comment by JanJereczek** [comment]
Mostly done with the modifications. I feel it would be nice to have a struct defining the parameters of the meta-analysis. Something like:

```julia
struct MetaAnalysisParameters
    n_surrogates::Int
    surrogate_method::Surrogate
    rng::AbstractRNG
    wv_indicator_width::Int
    wv_indicator_stride::Int
    wv_evolution_width::Int
    wv_evolution_stride::Int
end
```
That way, we can pass these parameters easier from a function to an other. What do you think?

**Comment by Datseris** [comment]
Sounds good!

**Comment by JanJereczek** [comment]
Not completely done yet but not missing much! I will write here when it is ready for review :)

**Comment by Datseris** [comment]
ok!

File: src/evolution_metrics.jl (hunk 2 above)
Comment (JanJereczek) [reply src/evolution_metrics.jl]: Correct, fixed!

File: src/evolution_metrics.jl (hunk 3 above)
Comment (JanJereczek) [reply src/evolution_metrics.jl]: solved by introducing the parameters struct

File: src/evolution_metrics.jl (hunk 4 above)
Comment (JanJereczek) [reply src/evolution_metrics.jl]: done!

File: src/evolution_metrics.jl (hunk 5 above)
Comment (JanJereczek) [reply src/evolution_metrics.jl]: good point, done

File: src/evolution_metrics.jl (hunk 6 above)
Comment (JanJereczek) [reply src/evolution_metrics.jl]: Next meeting could you explain when it makes sense to not inline functions?

File: src/evolution_metrics.jl (hunk 7 above)
Comment (JanJereczek) [reply src/evolution_metrics.jl:240]: done!

File: src/evolution_metrics.jl (hunk 8 above)
Comment (JanJereczek) [reply src/evolution_metrics.jl]: I should have it now, but not sure if you meant something different

File: src/significance.jl (hunk 9 above)
Comment (JanJereczek) [reply src/significance.jl]: corrected. also `i_lo` and `i_hi` are integers

**Comment by JanJereczek** [comment]
Ready for review! I think we should speak about naming in our next meeting. I already feel we are close to having a versatile interface :)

**Comment by Datseris** [comment]
I'm pushing some commits here to organize things a bit. Two general comments:

- Do not declare types unless there is a reason to. Every method you've written is cluttered with `::AbstractVector{T}`. Why is it important that `x` and `y` need to be the same type?
- The docs need some love. At the moment several docstrings are wrong or do not write all arguments correctly.
- There are still way more functions than necessary. I'll try to resolve this, have a look at the commits I will be pushing

**Comment by Datseris** [comment]
Actually, Jan, forget what I said. The code is fine the way it is now. Here is what I propose: make sure tests pass. Then, I merge and we start working on the documentation. The documentation will allow us to see what things are overengineered and which are not.

**Comment by JanJereczek** [comment]
Hi George,

Thanks for your corrections! Some remarks:
- I kept the precomputed version of computing the ridge slope and put a small benchmark experiment for you to check that it is way faster (test/metaanalysis_trend.jl)
- We seem to converge on following naming: we compute *indicators* and study their *evolution metrics*, which results in a *meta-analysis* of the indicators.
- Regarding the strong typing I require: I agree with you that it is not necessary but I felt like it was not requiring a lot and would potentially improve performance. But I guess we can just be less restrictive in that regard.
- I will make a small example on a time-series displaying a transition vs. one time-series without transition. This will be neat for the docs and will make sure the docstrings remain cleaner!

After merge into main I see following things to do next:
- A thresholding function shooting an alarm if `n_min_indicators` or more show a significant meta-analysis.
- Implement some more indicators.

What do you think?

**Comment by Datseris** [comment]
Highest priority now is the documentation, so that you can explain how you envision the package to be used. Then, we follow with your two suggestions.

**Pull request merged**